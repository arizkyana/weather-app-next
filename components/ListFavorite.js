import React from 'react'
import Link from "next/link";

class ListFavorite extends React.Component {
    constructor(props){
        super(props)
    }

    render() {
        const Favs = () => {
            if (!this.props.favs || this.props.favs.length <= 0) {
                return (
                    <div>
                        <small className="text-red-500">Favorite is empty</small>
                    </div>
                )
            } else {
                return this.props.favs.map((f, i) => (
                    <Link key={i} as={`/q/${f.city}`} href={`/?city=${f.city}`}>
                        <div
                            className="border mt-1 border-gray-500 px-3 py-3 cursor-pointer hover:bg-teal-500 hover:border-teal-500 hover:text-white">
                            <h1>{f.city}</h1>
                        </div>
                    </Link>
                ))
            }
        };

        return (
            <Favs/>
        )

    }
}

export default ListFavorite