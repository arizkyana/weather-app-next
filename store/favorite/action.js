const addToFavorite = (city) => {
  return {type: 'ADD_TO_FAVORITE', city}
};

export default {
    addToFavorite,
}