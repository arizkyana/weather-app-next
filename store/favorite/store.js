import { createStore } from "redux";
import _ from 'lodash'

import consola from 'consola'

function favorite(state = [], action){
    switch (action.type) {
        case 'ADD_TO_FAVORITE':
            state.push({city: action.city});
            return state;
    }
}

const store = createStore(favorite);

store.subscribe(() => consola.info('listen favorite state', store.getState()));

export default store