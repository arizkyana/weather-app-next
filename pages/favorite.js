import React from 'react'

import WithNavbar from '../layouts/WithNavbar'
import Header from '../components/Head'
import ListFavorite from '../components/ListFavorite'

import favorite from '../store/favorite'

import consola from 'consola'

class Favorite extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            favorites: []
        }
    }

    async componentDidMount() {
        const favorites = await favorite.store.getState();
        this.setState({favorites});

        consola.info('favorites', this.state.favorites)
    }

    render() {

        return (
            <WithNavbar className="px-6">
                <Header title="Favorites"/>
                <ListFavorite favs={this.state.favorites}/>
            </WithNavbar>
        )
    }
}

export default Favorite