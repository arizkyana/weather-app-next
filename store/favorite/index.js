import action from './action'
import store from './store'

export default {
    action,
    store
}