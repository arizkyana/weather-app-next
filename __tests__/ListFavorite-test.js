import React from 'react'
import {
    render,
    cleanup
} from '@testing-library/react'
import _ from 'lodash'
import {shallow, mount} from "enzyme";

import ListFavorite from '../components/ListFavorite'

import favorite from '../store/favorite'

import consola from 'consola'

afterEach(cleanup);

// expect list all favorite are listed
it('List all favorite are listed', () => {
    const favs = [
        {
            city: 'Bandung'
        },
        {
            city: 'Jakarta'
        },
        {
            city: 'Semarang'
        }
    ];
   const _listFavorite = render(
       <ListFavorite favs={favs}/>
   );



   expect(_listFavorite).toMatchSnapshot()


});

// expect not listed favorites
it('Empty favorites', () => {

    const _listFavorite = mount(
        <ListFavorite favs={[]}/>
    );

    expect(_listFavorite.find('small').text()).toBe("Favorite is empty")
});

// expect failed key favorites
it('Failed key favorites', () => {
    const favs = [
        {
            town: 'Bandung'
        },
        {
            town: 'Jakarta'
        },
        {
            town: 'Semarang'
        }
    ];
    const _listFavorite = shallow(
        <ListFavorite favs={favs}/>
    );

    expect(_listFavorite.children()).toMatchObject({})
});

// adding city to state
it('Adding city to state',  () => {
    favorite.store.dispatch(favorite.action.addToFavorite('Bandung'));
    const currentCity = favorite.store.getState();

    const findBandung = _.find(currentCity, {city: 'Bandung'});
    expect(findBandung).toMatchObject({city: "Bandung"})
});

// adding same city to state
it('Adding same city to state', () => {
    favorite.store.dispatch(favorite.action.addToFavorite('Bandung'));
    const currentCity = favorite.store.getState();

    const findBandung = _.find(currentCity, {city: 'Bandung'});

    expect(findBandung.city === 'Bandung').toBeTruthy()
});