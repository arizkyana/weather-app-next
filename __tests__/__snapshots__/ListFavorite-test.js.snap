// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List all favorite are listed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="border mt-1 border-gray-500 px-3 py-3 cursor-pointer hover:bg-teal-500 hover:border-teal-500 hover:text-white"
      >
        <h1>
          Bandung
        </h1>
      </div>
      <div
        class="border mt-1 border-gray-500 px-3 py-3 cursor-pointer hover:bg-teal-500 hover:border-teal-500 hover:text-white"
      >
        <h1>
          Jakarta
        </h1>
      </div>
      <div
        class="border mt-1 border-gray-500 px-3 py-3 cursor-pointer hover:bg-teal-500 hover:border-teal-500 hover:text-white"
      >
        <h1>
          Semarang
        </h1>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="border mt-1 border-gray-500 px-3 py-3 cursor-pointer hover:bg-teal-500 hover:border-teal-500 hover:text-white"
    >
      <h1>
        Bandung
      </h1>
    </div>
    <div
      class="border mt-1 border-gray-500 px-3 py-3 cursor-pointer hover:bg-teal-500 hover:border-teal-500 hover:text-white"
    >
      <h1>
        Jakarta
      </h1>
    </div>
    <div
      class="border mt-1 border-gray-500 px-3 py-3 cursor-pointer hover:bg-teal-500 hover:border-teal-500 hover:text-white"
    >
      <h1>
        Semarang
      </h1>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
